presto-spark-base/src/main/java/com/facebook/presto/spark/execution/LocalTaskFactory.java

```diff
@@ -0,0 +1,107 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.execution;
+
+import com.facebook.airlift.concurrent.BoundedExecutor;
+import com.facebook.airlift.http.client.HttpClient;
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.TaskInfo;
+import com.facebook.presto.execution.TaskSource;
+import com.facebook.presto.execution.buffer.OutputBuffers;
+import com.facebook.presto.execution.scheduler.TableWriteInfo;
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.server.TaskUpdateRequest;
+import com.facebook.presto.server.remotetask.RemoteTaskStats;
+import com.facebook.presto.sql.planner.PlanFragment;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class LocalTaskFactory
+{
+    private final HttpClient httpClient;
+    private final RemoteTaskStats stats;
+    private final ExecutorService coreExecutor;
+    private final Executor executor;
+    private final ScheduledExecutorService updateScheduledExecutor;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    @Inject
+    public LocalTaskFactory(
+            @ForWorker HttpClient httpClient,
+            RemoteTaskStats stats,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskInfo> taskInfoCodec)
+    {
+        this.httpClient = httpClient;
+        this.stats = stats;
+        this.taskUpdateRequestCodec = taskUpdateRequestCodec;
+        this.planFragmentCodec = planFragmentCodec;
+        this.taskInfoCodec = taskInfoCodec;
+
+        this.coreExecutor = newCachedThreadPool(daemonThreadsNamed("local-task-callback-%s"));
+        this.executor = new BoundedExecutor(coreExecutor, 1000);
+        this.updateScheduledExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("task-info-scheduler-%s"));
+    }
+
+    public LocalTask createLocalTask(
+            Session session,
```

Comment:
Session is per query basis to override some of the worker configs. For batch the unit of work during a worker process lifetime is per task which already surpasses a query lifetime. We can solely rely on worker config. In this case we do not want to expose session to the caller.

Another possibility would be there is already plumbings in Sapphire to pass session from the top (client->driver->executor) then we can just leave it as is.

Reply:
Currently, the session is used when creating TaskUpdateRequest. Let me know if I should use some other fields to get the session and extraCredentials for the TaskUpdateRequest.

new TaskUpdateRequest(
                session.toSessionRepresentation(),
                session.getIdentity().getExtraCredentials(),
                fragment,
                sources,
                outputBuffers.get(),
                writeInfo);

Reply:
Right, I think we need to figure out if current Sapphire uses any sessions at all. 

If not then having Session here does not make sense because caller is not aware of any session related stuff. One way is to remove Session exposure from this API and create TaskUpdateRequest with some dummy Session created on-site.

If yes then we will need to leave Session here in the API. I think unlikely a session would be needed here though.